builder/frameworks/arduino.py

```
@@ -28,250 +28,574 @@
-    packages_to_install = []
-    for package, spec in deps.items():
-        if package not in installed_packages:
-            packages_to_install.append(package)
-        else:
-            version_spec = semantic_version.Spec(spec)
-            if not version_spec.match(installed_packages[package]):
-                packages_to_install.append(package)
+    packages_to_install = list(get_packages_to_install(python_deps, installed_packages))
 
     if packages_to_install:
+        packages_str = " ".join(f'"{p}{deps[p]}"' for p in packages_to_install)
```

Ruff (0.11.9)

Undefined name `deps`

(F821)